app/src/main/java/de/westnordost/streetcomplete/screens/user/profile/ProfileFragment.kt

```diff
@@ -114,33 +52,60 @@ class ProfileFragment : Fragment(R.layout.fragment_profile) {
         binding.currentWeekLocalRankText.background = LaurelWreathDrawable(resources)
         binding.currentWeekGlobalRankText.background = LaurelWreathDrawable(resources)
 
-        binding.logoutButton.setOnClickListener {
-            userLoginStatusController.logOut()
-        }
+        binding.logoutButton.setOnClickListener { viewModel.logOutUser() }
         binding.profileButton.setOnClickListener {
-            openUri("https://www.openstreetmap.org/user/" + userDataSource.userName)
+            openUri("https://www.openstreetmap.org/user/" + viewModel.userName.value)
         }
-    }
-
-    override fun onStart() {
-        super.onStart()
 
-        viewLifecycleScope.launch {
-            userDataSource.addListener(userListener)
-            userUpdater.addUserAvatarListener(userAvatarListener)
-            statisticsSource.addListener(statisticsListener)
-            unsyncedChangesCountSource.addListener(unsyncedChangesCountListener)
-            achievementsSource.addListener(achievementsListener)
-
-            updateUserName()
-            updateAvatar()
-            updateEditCountTexts()
-            updateUnpublishedEditsText()
-            updateDaysActiveText()
-            updateGlobalRankTexts()
-            updateLocalRankTexts()
-            updateAchievementLevelsText()
-            updateDatesActiveView()
+        observe(viewModel.userName) { name ->
+            binding.userNameTextView.text = name
+        }
+        observe(viewModel.userAvatarFile) { file ->
+            val avatar = if (file.exists()) BitmapFactory.decodeFile(file.path) else anonAvatar
+            binding.userAvatarImageView.setImageBitmap(avatar)
+        }
+        observe(viewModel.editCount) { count ->
+            binding.editCountText.text = count.toString()
+        }
+        observe(viewModel.editCountCurrentWeek) { count ->
+            binding.currentWeekEditCountText.text = count.toString()
+        }
+        observe(viewModel.achievementLevels) { levels ->
+            binding.achievementLevelsContainer.isGone = levels <= 0
+            binding.achievementLevelsText.text = levels.toString()
+            binding.achievementLevelsText.background.level = min(levels / 2, 100) * 100
```

Comment:
calculations done in UI, slow downs. Better to be done in view model via a custom data object and flows.